activesupport/lib/active_support/cache/redis_cache_store.rb

```
@@ -0,0 +1,389 @@
+        def write_multi_entries(entries, options)
+          if @mset_capable && options.none?
+            begin
+              write_multi_entries_mset(entries, options)
+            rescue Redis::Distributed::CannotDistribute
+              @mset_capable = false
+              super
+            end
+          else
+            super
+          end
+        end
+
+        def write_multi_entries_mset(entries)
+          if entries.any?
+            failsafe :write_entries do
+              redis.mapped_mset(entries)
+            end
+          end
+        end
+
+        # Truncate keys that exceed 1kB.
+        def normalize_key(key, options)
+          truncate_key super
+        end
+
+        def truncate_key(key)
+          if key.bytesize > max_key_bytesize
+            suffix = ":md5:#{Digest::MD5.hexdigest(key)}"
+            truncate_at = max_key_bytesize - suffix.bytesize
+            "#{key.byteslice(0, truncate_at)}#{suffix}"
+          else
+            key
+          end
+        end
+
+        def deserialize_entry(raw_value)
+          if raw_value
+            entry = Marshal.load(raw_value) rescue raw_value
+            entry.is_a?(Entry) ? entry : Entry.new(entry)
+          end
+        end
+
+        def serialize_entry(entry)
+          Marshal.dump(entry)
+        end
+
+        def failsafe(method, returning: nil)
+          yield
+        rescue ::Redis::BaseConnectionError => e
+          handle_exception exception: e, method: method, returning: returning
+          returning
+        end
+
+        def handle_exception(exception:, method:, returning:)
+          if @error_handler
+            @error_handler.(method: method, exception: exception, returning: returning)
+          end
+        rescue => failsafe
+          warn "RedisCacheStore ignored exception in handle_exception: #{failsafe.class}: #{failsafe.message}\n  #{failsafe.backtrace.join("\n  ")}"
+        end
```

Exceptions are swallowed by default. Apps provide an `error_handler` to report on them.